### DISCUSSION TITLE
Rust panic on misuse

### DISCUSSION THREAD

**Discussion by forty** [jedisct1/libsodium#1352] [General]
I'm using libsodium from rust via https://github.com/jedisct1/libsodium-sys-stable and I figured that it would be nice to panic on misuse, which is convenient for example for testing

```rust
#[no_mangle]
extern "C" fn misuse_panic() {
    panic!("sodium_misuse");
}

unsafe {
    sodium_set_misuse_handler(Some(misuse_panic));
}
```

This is convenient for testing, for example
```rust
#[test]
#[should_panic]
fn test_misuse() {
    unsafe {
        sodium_init();
        sodium_misuse();
    }
}
```

Issue is that sodium_misuse set the internal sodium lock and then never unlock it https://github.com/jedisct1/libsodium/blob/master/src/libsodium/sodium/core.c#L196 which ends up dead locking my tests

Can you think of a good way to handle this? Or is this just all too horrible anyway, and I should just let the process crash on misuse? :)

Thanks!

**Comment by jedisct1** [comment]
Let it crash. `misuse` should never be called, unless there's a ginormous bug in the application or in libsodium itself. In which case, immediately aborting is the best thing to do. You should especially not use libsodium again after this happens, so keeping the lock is not a bad thing.
